pkg/controller/utils/datadog/forwarders_manager.go

```diff
@@ -0,0 +1,83 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2019 Datadog, Inc.
+
+package datadog
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ForwardersManager is a collection of metricsForwarder per DatadogAgentDeployment
+// ForwardersManager implements the controller-runtime Runnable interface
+type ForwardersManager struct {
+	forwarders map[string]*metricsForwarder
+	k8sClient  client.Client
+	sync.Mutex
+}
+
+// NewForwardersManager builds a new ForwardersManager
+// ForwardersManager implements the controller-runtime Runnable interface
+func NewForwardersManager(k8sClient client.Client) *ForwardersManager {
+	return &ForwardersManager{
+		k8sClient:  k8sClient,
+		forwarders: make(map[string]*metricsForwarder),
+	}
+}
+
+// Start must be handled by the controller-runtime manager
+func (f *ForwardersManager) Start(stop <-chan struct{}) error {
+	<-stop
+	f.stopAllForwarders()
+	return nil
+}
+
+// Register starts a new metricsForwarder if a new DatadogAgentDeployment is detected
+func (f *ForwardersManager) Register(namespacedName types.NamespacedName) {
+	f.Lock()
+	defer f.Unlock()
+	id := namespacedName.String()
+	if _, found := f.forwarders[id]; !found {
+		log.Info("New Datadog metrics forwarder registred", "ID", id)
+		f.forwarders[id] = newMetricsForwarder(f.k8sClient, namespacedName)
+		go f.forwarders[id].start()
+	}
+}
+
+// Unregister stops a metricsForwarder when its corresponding DatadogAgentDeployment is deleted
+func (f *ForwardersManager) Unregister(namespacedName types.NamespacedName) {
+	id := namespacedName.String()
+	log.Info("Unregistering metrics forwarder", "ID", id)
+	if err := f.unregisterForwarder(id); err != nil {
+		log.Error(err, "cannot unregister metrics forwarder", "ID", id)
+		return
+	}
+	log.Info("Metrics forwarder unregistered", "ID", id)
+}
+
+// stopAllForwarders stops the running metricsForwarder goroutines
+func (f *ForwardersManager) stopAllForwarders() {
+	f.Lock()
+	defer f.Unlock()
+	for id, forwarder := range f.forwarders {
+		log.Info("Stopping metrics forwarder", "ID", id)
+		forwarder.stop()
```

Comment:
good catch, actually the `stop()` only closes the stop channel (non blocking), so I added a wait group so that the forwarder `start()` method decrements it before it returns.